File: ts/384/287.ts
// 二分查找
function findDuplicate(nums: number[]): number {
  const n = nums.length
  let l = 1, r = n - 1, ans = -1
  while (l <= r) {
    let m = l + ((r - l) >> 1)
    let cnt = 0
    for (let i = 0; i < n; ++i) {
      cnt += nums[i] <= m ? 1 : 0
    }
    if (cnt <= m) l = m + 1
    else {
      r = m - 1
      ans = m
    }
  }
  return ans
};

File: py3/384/287.py
from typing import List

class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        n = len(nums)
        l, r, ans = 1, n - 1, -1
        while l <= r:
            m = l + (r - l) // 2
            cnt = 0
            for v in nums:
                cnt += 1 if v <= m else 0

            if cnt <= m:
                l = m + 1
            else:
                r = m - 1
                ans = m
        return ans